### PR TITLE
Fix stream ordering in ORC writer

### PR DESCRIPTION
**Pull request by sdruzkin**
## Description
Fix stream ordering in the ORC writer:

For default order:
1. Place steams that belong to the same ~~node~~ column together.
2. Order columns by ~~node~~ column size in desc order.
3. Order streams in the column by node id to preserve complex type tree iteration (reading) order.
4. Group flatmap streams by sequence, order groups inside of the column by the size of "column + sequence" group in desc order. 

For predefined (feature) order:
1. Order streams by node and kind inside of the column+sequence group for feature ordered streams.

## Motivation and Context
Old implementation could scatter streams that belong to the same column across
the stripe instead of grouping them together, this led to poor reader IO.

## Impact
With better grouping and ordering read IO for complex types should be better.

## Test Plan
Added new unit tests.
Verifier tests:
```
Progress: 9263 succeeded, 544 skipped, 0 resolved, 3 failed, 100.00% done
```

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/StreamDataOutput.java [hunk 1]
```diff
@@ -68,4 +69,12 @@ public void writeData(SliceOutput sliceOutput)
         long size = writer.applyAsLong(sliceOutput);
         verify(stream.getLength() == size, "Data stream did not write expected size");
     }
+
+    @Override
+    public String toString()
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/StreamDataOutput.java:74]: Added toString to get nice error messages in tests

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/StreamOrderingLayout.java [hunk 2]
```diff
@@ -33,6 +35,21 @@
 public class StreamOrderingLayout
         implements StreamLayout
 {
+    private static final Comparator<StreamDataOutput> IN_GROUP_COMPARATOR = (streamDataA, streamDataB) -> {
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/StreamOrderingLayout.java:38]: This comparator is used to sort streams inside of the flat map sequence tree.

**Comment by sdruzkin** [comment]
Unrelated tests failed :(

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java [hunk 3]
```diff
@@ -130,4 +132,28 @@ public Stream withOffset(long offset)
...
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Stream stream = (Stream) o;
+        return column == stream.column
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java:146]: Is column always properly populated? I vaguely remembered it was not used

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnSizeLayout.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+            // sort by the stream kind, we don't really need it, but it makes testing easier
+            return Integer.compare(streamA.getStreamKind().ordinal(), streamB.getStreamKind().ordinal());
+        });
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ByColumnSize{}";
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnSizeLayout.java:146]: Fix name :)

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java (hunk 3 above)
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java:146]: It's actually a misnamed nodeId, streams in presto-orc don't have the column.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnSizeLayout.java (hunk 4 above)
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnSizeLayout.java]: Fixed 👍

**Comment by zzhao0** [comment]
LGTM!

**Comment by sdruzkin** [comment]
> LGTM!

@NikhilCollooru could you please take a look and stamp this diff if everything looks good?

**Review by NikhilCollooru — APPROVED** [review]

**Pull request merged**